### PR TITLE
Fix compiler.links for vala

### PR DESCRIPTION
**Pull request by bcorby**
Fixes issue #12959
compiler.links command for vala crashes

### CODE REVIEW

File: mesonbuild/compilers/vala.py [hunk 1]
```diff
@@ -41,6 +46,22 @@ def get_output_args(self, outputname: str) -> T.List[str]:
...
+        args: T.List[str] = []
+        if mode is CompileCheckMode.LINK and self.force_link:
+            return args
+        args += self.get_always_args()
+        if mode is CompileCheckMode.COMPILE:
+            args += self.get_compile_only_args()
+        elif mode is CompileCheckMode.PREPROCESS:
+            args += self.get_preprocess_only_args()
+        else:
+            assert mode is CompileCheckMode.LINK
```
Comment (dcbaker) [inline mesonbuild/compilers/vala.py:59]: I don't think this is needed, the type checker should enforce this for us.

File: mesonbuild/compilers/vala.py [hunk 2]
```diff
@@ -119,3 +140,24 @@ def thread_flags(self, env: 'Environment') -> T.List[str]:
...
+
+    def get_option_link_args(self, options: 'KeyedOptionDictType') -> T.List[str]:
+        return []
+
+    def links(self, code: 'mesonlib.FileOrString', env: 'Environment', *,
+              compiler: T.Optional['Compiler'] = None,
+              extra_args: T.Union[None, T.List[str], CompilerArgs, T.Callable[[CompileCheckMode], T.List[str]]] = None,
+              dependencies: T.Optional[T.List['Dependency']] = None,
+              disable_cache: bool = False) -> T.Tuple[bool, bool]:
+        self.force_link = True
```
Comment (dcbaker) [inline mesonbuild/compilers/vala.py:152]: I don't love this, but I don't see a better way to handle this problem :/

**Review by dcbaker — CHANGES_REQUESTED** [review]

File: mesonbuild/compilers/vala.py (hunk 1 above)
Comment (bcorby) [reply mesonbuild/compilers/vala.py]: Agreed.

File: mesonbuild/compilers/vala.py (hunk 2 above)
Comment (bcorby) [reply mesonbuild/compilers/vala.py:150]: Yep, me too.

**Review by dcbaker — APPROVED** [review]

**Comment by bcorby** [comment]
No problem. Will this be in 1.4.1?

**Comment by dcbaker** [comment]
I added it to the milestone, I think that means it will automatically be picked up

**Comment by bcorby** [comment]
Thanks.

**Pull request merged**